action/penalization/allegation.go

```diff
@@ -0,0 +1,162 @@
+func (r Allegation) Signers() []action.Address {
+	return []action.Address{r.ValidatorAddress.Bytes()}
+}
+
+func (r Allegation) Type() action.Type {
+	return action.ALLEGATION
+}
+
+func (r Allegation) Tags() kv.Pairs {
+	tags := make([]kv.Pair, 0)
+
+	tag := kv.Pair{
+		Key:   []byte("tx.type"),
+		Value: []byte(r.Type().String()),
+	}
+	tag2 := kv.Pair{
+		Key:   []byte("tx.validator"),
+		Value: r.ValidatorAddress.Bytes(),
+	}
+	tag3 := kv.Pair{
+		Key:   []byte("tx.malicious"),
+		Value: r.MaliciousAddress.Bytes(),
+	}
+
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, uint64(r.BlockHeight))
+	tag4 := kv.Pair{
+		Key:   []byte("tx.height"),
+		Value: b,
+	}
+	tag5 := kv.Pair{
+		Key:   []byte("tx.proof"),
+		Value: []byte(r.ProofMsg),
+	}
+	tag6 := kv.Pair{
+		Key:   []byte("tx.requestID"),
+		Value: []byte(r.RequestID),
+	}
+
+	tags = append(tags, tag, tag2, tag3, tag4, tag5, tag6)
+	return tags
+}
+
+var _ action.Tx = allegationTx{}
+
+type allegationTx struct{}
+
+func (atx allegationTx) Validate(ctx *action.Context, tx action.SignedTx) (bool, error) {
+	r := &Allegation{}
+	err := r.Unmarshal(tx.Data)
+	if err != nil {
+		return false, errors.Wrap(action.ErrWrongTxType, err.Error())
+	}
+	err = action.ValidateBasic(tx.RawBytes(), r.Signers(), tx.Signatures)
+	if err != nil {
+		return false, err
+	}
+
+	if err := r.ValidatorAddress.Err(); err != nil {
+		return false, err
+	}
+
+	if err := r.MaliciousAddress.Err(); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (atx allegationTx) ProcessCheck(ctx *action.Context, tx action.RawTx) (ok bool, result action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' transaction for ProcessCheck", tx)
+	ok, result = runAllegationTransaction(ctx, tx)
+	ctx.Logger.Debug("Result 'allegation' transaction for ProcessCheck", ok, result)
+	return
+}
+
+func (atx allegationTx) ProcessDeliver(ctx *action.Context, tx action.RawTx) (ok bool, result action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' transaction for ProcessDeliver", tx)
+	ok, result = runAllegationTransaction(ctx, tx)
+	ctx.Logger.Debug("Result 'allegation' transaction for ProcessDeliver", ok, result)
+	return
+}
+
+func (atx allegationTx) ProcessFee(ctx *action.Context, signedTx action.SignedTx, start action.Gas, size action.Gas) (bool, action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' Transaction for ProcessFee", signedTx)
+	r := &Allegation{}
+	err := r.Unmarshal(signedTx.Data)
+	if err != nil {
+		return false, action.Response{Log: errors.Wrap(err, "failed to unmarshal").Error()}
+	}
+	return action.StakingPayerFeeHandling(ctx, r.ValidatorAddress, signedTx, start, size, 1)
+}
+
+func runAllegationTransaction(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	al := &Allegation{}
+	err := al.Unmarshal(tx.Data)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, action.ErrWrongTxType, al.Tags(), err)
+	}
+
+	feeOpt, err := ctx.GovernanceStore.GetFeeOption()
```

Comment:
I think we should still keep it in validate , because validate is run during the broadcast as well .
Instead we can just replace the access from the gov store by the the context object ctx.FeePool.GetOpt()